key_gen.go

```diff
@@ -57,17 +79,14 @@ func KeyForPrefixedStringMap(
 	return keyForPrefixedStringMaps(prefix, stringMap)
 }
 
-// keyForPrefixedStringMaps generates a unique key for a prefix and a series
-// of maps containing tags.
-//
-// If a key occurs in multiple maps, keys on the right take precedence.
-func keyForPrefixedStringMaps(prefix string, maps ...map[string]string) string {
-	keys := keyGenPool.stringsPool.Get().([]string)
+func keyForPrefixedStringMapsAsKey(prefix string, maps ...map[string]string) key {
+	keys := keyGenPool.stringsPool.Get()
```

Comment:
```suggestion
	// fixed size slice will be stack allocated
	keys := make([]string, 0, 16)
```

Instead of a using a object-pool, could you use a stack-allocated string slice as in my suggestion? The initial size is e.g., 16 and if there are more than 16 keys, the `append` in line 86 will extend it (and thus causing an allocation). The good thing is that you don't have to remember to clean up the slice afterward and it is inherently thread-safe.

Reply:
It is a good suggestion for keys as the string slice is not returned from the method. However unfortunately the stack allocated string slice escapes to heap as the sort method is called.

Reply:
to avoid this in `m3collector` we actually use a simple insertion sort as it's better than an alloc+virtualized call as tag count is bounded by limits on metric size.

Reply:
I have removed the alloc with sorting by using a pointer to sort.StringSlice and using the sort.Sort() method.

Reply:
Ahh sorry, just noticed your comment on the heap-allocation. 

For reference, diving even deeper into `sort.Sort` (which is called by `sort.Strings`), it seems that the escape analysis has determined that parameters to `sort.Sort` escapes:
```
./key_gen.go:61:2: keys escapes to heap:
./key_gen.go:61:2:   flow: {heap} = &keys:
./key_gen.go:61:2:     from &keys (address-of) at ./key_gen.go:69:32
./key_gen.go:61:2:     from (*sort.StringSlice)(&keys) (interface-converted) at ./key_gen.go:69:31
./key_gen.go:61:2:     from sort.Sort((*sort.StringSlice)(&keys)) (call parameter) at ./key_gen.go:69:11
```
as far as I can understand the `(call parameter)` means that the parameter escapes. Thus there is no way to call `sort.Sort` with a stack-allocated value.

Reply:
thanks. I have removed heap allocated []string and also used insertion sort (copied from go [sort](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/sort/sort.go;l=38) pkg)